### PR TITLE
Dust emission

### PR DESCRIPTION
**Pull request by stephenmwilkins**
## Issue Type

- [ ] Bug
- [ ] Document
- [x] Enhancement

## Checklist

- [x] I have added docstrings to all methods
- [x] I have added sufficient comments to all lines
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no pep8 errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Added dust emission model as well as methods to generate dust emission spectrum using energy balance approach. Examples of both the dust emission model and a parametric model now implemented in docs. 

NOTE: because the test grid is truncated at 20um this example doesn't work particularly well at the moment. I will have to extend the test grid but there is no point doing that until the grid_update PR is accepted.

### CODE REVIEW

**Comment by stephenmwilkins** [comment]
I have merged `grid_update` branch and also generated a new test_grid with log10 wavelength spacing extending into the FIR.

**Comment by stephenmwilkins** [comment]
This shouldn't be properly reviewed until the `grid_test` PR is approved since it has been merged with that.

**Comment by christopherlovell** [comment]
Needs updating with main, looks like there are conflicts

File: docs/source/misc/dust_emission.ipynb
Comment (christopherlovell) [inline docs/source/misc/dust_emission.ipynb:1]: Why is this in a separate 'misc' folder? make a new dedicated dust emission section

**Review by christopherlovell — COMMENTED** [review]
Few minor edits required. Have not tested functionality explicitly

File: synthesizer/dust.py [hunk 1]
```diff
@@ -313,3 +313,229 @@ def attenuate(self, tau_V, lam):
...
+
+
+
+from scipy import integrate
+from unyt import h, c, kb, um, erg, s, Hz
+from unyt import accepts, returns
+from unyt.dimensions import length, time, temperature
+
+
+def planck(nu, T):
```
Comment (christopherlovell) [inline synthesizer/dust.py:327]: Could this be moved to `utils.py`?

Comment (WillJRoper) [reply synthesizer/dust.py]: Have moved

**Comment by WillJRoper** [comment]
I've done some linting in `dust.py` but left the rest for @stephenmwilkins. Lots of cleaning is needed, whitespace around operators, removal of blank lines, and `power_law` needs to be `PowerLaw`, to name a few issues. 

install flake8 and run `flake8 <path/to/file>` on the terminal to see a list of issues.

I'm also somewhat confused by the classes in `dust.py`. None of them have an init so none have attributes. They are essentially a container of functions. This should probably be in files of functions if that is all that is needed.

**Comment by christopherlovell** [comment]
> I'm also somewhat confused by the classes in `dust.py`. None of them have an init so none have attributes. They are essentially a container of functions. This should probably be in files of functions if that is all that is needed.

All the emission derived classes have `__init__`'s with attributes? Or are you talking about the dust curves?

As an aside, I'd argue the emission classes should be moved to their own dedicated script `dust_emission.py`

**Comment by WillJRoper** [comment]
My bad. I somehow missed that the children do. It's the parent that doesn't but it should and the children should instantiate it. 

Agreed on the split of the file.

**Comment by stephenmwilkins** [comment]
> My bad. I somehow missed that the children do. It's the parent that doesn't but it should and the children should instantiate it.
> 
> Agreed on the split of the file.

Yes, we either need `dust_attenuation.py` and `dust_emission.py` or `dust/emission.py` and `dust/attenuation.py`. Not sure what's best?

**Comment by WillJRoper** [comment]
The latter gets my vote. Submodules give a nice structure.

**Pull request merged**